include/seqan3/search/kmer_index/shape_iterator.hpp

```diff
@@ -0,0 +1,497 @@
+                                                 std::conditional_t<std::BidirectionalIterator<it_t>,
+                                                                    typename std::bidirectional_iterator_tag,
+                                                                    typename std::forward_iterator_tag>>;
+
+private:
+    //!\brief The size type of this iterator.
+    using position_type = std::make_unsigned_t<difference_type>;
+    //!\brief The alphabet type of the passed iterator.
+    using alphabet_t = value_type_t<it_t>;
+
+    //!\brief The alphabet size.
+    static constexpr size_t const sigma{alphabet_size_v<alphabet_t>};
+
+    //!\brief The hash value.
+    size_t hash_value{0};
+
+    //!\brief The seqan3::shape to use for hashing.
+    shape shape_;
+
+    //!\brief The factor for the left most position of the hash value.
+    size_t roll_factor{0};
+
+    //!\brief True if `shape_` contains only 1 (ungapped), false otherwise.
+    bool ungapped_shape{false};
+
+    //!\brief Iterator pointing to begin of the underlying text.
+    it_t text_start;
+
+    //!\brief Iterator to the leftmost position of the k-mer.
+    it_t text_left;
+
+    //!\brief Iterator to the rightmost position of the k-mer.
+    it_t text_right;
+
+    //!\brief Increments iterator by 1.
+    void hash_forward()
+    {
+        if (ungapped_shape)
+        {
+            hash_roll();
+        }
+        else
+        {
+            std::advance(text_left,  1);
+            hash_full();
+        }
+    }
+
+    /*!\brief Increments iterator by `skip`.
+     * \param skip Amount to increment.
+     * \attention This function is only avaible if it_t models std::RandomAccessIterator.
+     */
+    void hash_forward(difference_type const skip)
+        //!\cond
+        requires std::RandomAccessIterator<it_t>
+        //!\endcond
+    {
+        std::advance(text_left, skip);
+        hash_full();
+    }
+
+    /*!\brief Decrements iterator by 1.
+     * \attention This function is only avaible if it_t models std::BidirectionalIterator.
+     */
+    void hash_reverse()
+        //!\cond
+        requires std::BidirectionalIterator<it_t>
+        //!\endcond
+    {
+        if (ungapped_shape)
+        {
+            hash_roll_back();
+        }
+        else
+        {
+            std::advance(text_left,  -1);
+            hash_full();
+        }
+    }
+
+    /*!\brief Decrements iterator by `skip`.
+     * \param skip Amount to decrement.
+     * \attention This function is only avaible if it_t models std::BidirectionalIterator.
+     */
+    void hash_reverse(difference_type const skip)
+    {
+        std::advance(text_left, -skip);
+        hash_full();
+    }
+
+    //!\brief Calculates a hash value by explicitly looking at each position.
+    void hash_full()
+    {
+        text_right = text_left;
+        hash_value = shape_[0] * to_rank(*text_right);
```

Comment:
what if the first position is "not inlcuded"? (it doesn't make much sense since you can just make your shape smaller by one... but we are also not enforcing/asserting this "studpidity")

Reply:
Mhm, I think it doesn't matter for the computation...it's still correct, just less efficient - in the worst case you have something like `0111` where you could just start computing the hashes one position later and use the `111` shape.
I guess we could add an assert to the constructor? AFAIK we assert in seqan2 that a shape ends and starts with a 1.

Reply:
Your right it's correct anyway, I didn't see that. I think asserts don't hurt :)